#TODO URL dashboard:

# Perkenalan

Nama : Jason Rich Darmawan Onggo Putra

Batch : 016 RMT

# Identifikasi Masalah

## Topik Permasalahan

Perbandingan Pembelian Alkohol di Iowa, Amerika Serikat sebelum Pandemi Covid-19 dan puncak (kasus baru per hari) Pandemi Covid-19

Data sebelum Pandemi Covid-19 yang digunakan adalah bulan Janauri 2018.

Data Puncak Pandemi Covid-19 yang digunakan adalah bulan Januari 2021.

## Problem Statement

Membuktikan Dampak (Meningkat/Berkurang/Tidak Berdampak) Pandemi Covid-19 terhadap Konsumsi Alkohol di Iowa, Amerika Serikat.

## Latar Belakang

Sebagai Data Analyst yang mungkin akan bekerja di e-commerce atau franchisor convenience store, saya perlu intuisi untuk "apa perlu menambah stok minuman beralkohol ketika terjadi krisis seperti Pandemi Covid-19"

## Penjabaran Masalah

#TODO penjabaran masalah

### Hypothesis Testing 1

- [ ] Apa rata-rata penjualan minuman beralkohol (dalam volume / column `volume_sold_liters`) di Iowa, Amerika Serikat pada bulan Januari 2018 dan 2021 sama/berbeda (lebih kecil/lebih besar)?
- [ ] Apa rata-rata penjualan minuman beralkohol (dalam total harga / column `sale_dollars`) di Iowa, Amerika Serikat pada bulan Januari 2018 dan 2021 sama/berbeda?

### Visualisasi

- [ ] Di kota (column `city`) mana penjualan tertinggi (dalam volume / column `volume_sold_liters`) minuman beralkohol di Iowa, Amerika Serikat pada bulan Januari 2018 dan 2021?
- [ ] Di kota (column `city`) mana penjualan tertinggi (dalam total harga / column `sale_dollars`) di Iowa, Amerika Serikat pada bulan Januari 2018 dan 2021?

### Hypothesis Testing 2

- [ ] Apa rata-rata harga minuman beralkohol (column `state_bottle_retail`) di Iowa, Amerika Serikat pada bulan Januari 2018 dan 2021 sama/berbeda?

### Dugaan Penyebab Perbedaan Rata-rata Penjualan Minuman Beralkohol pada bulan Januari 2018 dan 2021

- [ ] Apa korelasi harga minimuman beralkohol (column `state_bottle_retail`) dengan jumlah botol terjual (column `bottles_sold`) di Iowa Amerika Serikat pada bulan Januari 2018 dan 2021?
- [ ] Apa minuman beralkohol dijual (column `state_bottle_retail`) dibawah biaya produksi (`state_bottle_cost`) di Iowa Amerika Serikat pada bulan Januari 2018 dan 2021?

### Informasi Lainnya

- [ ] Apa ukuran botol (column `bottle_volume_ml`) yang paling banyak terjual (column `bottles_sold`) di Iowa, Amerika Serikat pada bulan Januari 2018 dan 2021?
- [ ] Apa kategori minuman beralkohol yang paling banyak terjual (column `bottles_sold`) di Iowa, Amerika Serikat pada bulan Januari 2018 dan 2021?

# Data Loading

## Import Library

In [13]:
import pandas as pd

## Reference

project: bigquery-public-data

dataset: iowa_liquor_sales

table: sales

## Query

advantage: sub-query agar lebih clean dan lebih mudah ganti tahun dan bulan.

disadvantage: result sub-query-nya boros (diload hanya untuk difilter di query berikutnya).

```
SELECT *,
FROM (
  SELECT *, 
         EXTRACT(YEAR FROM date) as year,
         EXTRACT(MONTH FROM date) as month
  FROM `bigquery-public-data.iowa_liquor_sales.sales`
  WHERE date BETWEEN "2018-01-01" AND "2022-01-31"
)
WHERE year IN (2018, 2022) 
 AND month=1
```

In [14]:
df = pd.read_csv("./bigquery-public-data_iowa_liquor_sales_year_2018_and_year_2022_month_1.csv")

## DataFrame.head()

#TODO drop columns
- [x] can't be visualized nor answer 'Penjabaran Masalah'
  `store_number`, `store_name`, `address`, `store_location`, `vendor_number`, `vendor_name`, `item_number`, `item_description`
- too big to derive conclusion from
  `county_number`, `county`
- duplicated columns
  `volume_sold_gallons`, `year`, `month`

#TODO DataFrame.sort_values() 
- [x] by column `date`

In [15]:
pd.options.display.max_columns = None
df.head()

invoice_and_item_number        date  store_number  \
0         INV-10095200050  2018-01-30          4167   
1         INV-09780600017  2018-01-15          4669   
2         INV-43674300036  2022-01-11          4843   
3         INV-43538900011  2022-01-05          5750   
4         INV-43515200004  2022-01-04          9047   

                             store_name              address        city  \
0              Iowa Street Market, Inc.         1256 Iowa St     Dubuque   
1                 Vom Fass / Des Moines         833, 42nd St  Des Moines   
2               Stewart Road Fast Break      2418 Stewart Rd   Muscatine   
3  Casey's General Store #2920 / Ankeny  1010 SE National Dr      Ankeny   
4           Revelton Distilling Company       1400 W Clay St     Osceola   

   zip_code                        store_location  county_number     county  \
0   52001.0          POINT (-90.668138 42.504959)           31.0    DUBUQUE   
1   50312.0  POINT (-93.674384 41.59343100000001)           77.0       POLK   
2   52761.0          POINT (-91.072733 41.390826)           70.0  MUSCATINE   
3   50021.0  POINT (-93.582179 41.70486400000001)           77.0       POLK   
4   50213.0          POINT (-93.787233 41.041472)           20.0     CLARKE   

    category                        category_name  vendor_number  \
0  1092100.0  Imported Distilled Spirit Specialty           55.0   
1  1901200.0                  Special Order Items          885.0   
2  1091300.0       Neutral Grain Spirits Flavored          346.0   
3  1091300.0       Neutral Grain Spirits Flavored          346.0   
4  1011700.0                        Corn Whiskies          654.0   

                   vendor_name  item_number  \
0        SAZERAC NORTH AMERICA        76227   
1    YAHARA BAY DISTILLERS INC       905081   
2     OLE SMOKY DISTILLERY LLC        86739   
3     OLE SMOKY DISTILLERY LLC       986845   
4  Revelton Distilling Company        26294   

                           item_description  pack  bottle_volume_ml  \
0                            Montezuma Blue    12              1000   
1         Yahara Bay Apple Crisp(Apple Pie)     6               200   
2  Ole Smoky Apple Pie Moonshine 70prf Mini     8                50   
3  SOOH Ole Smoky Blackberry Moonshine Mini     8                50   
4         Revelton Shine White Whiskey Mini     6                50   

   state_bottle_cost  state_bottle_retail  bottles_sold  sale_dollars  \
0               4.66                 6.99            60        419.40   
1               7.98                11.97             6         71.82   
2               8.75                13.13             3         39.39   
3               8.75                13.13             4         52.52   
4              14.40                21.60             6        129.60   

   volume_sold_liters  volume_sold_gallons  year  month  
0               60.00                15.85  2018      1  
1                1.20                 0.32  2018      1  
2                0.15                 0.03  2022      1  
3                0.20                 0.05  2022      1  
4                0.30                 0.07  2022      1

In [16]:
# Is column invoice_and_item_number value unique?
df.loc[df['invoice_and_item_number'].map(df['invoice_and_item_number'].value_counts()) > 1]
# alternative method:
# df.loc[df['invoice_and_item_number'].duplicated()]

Empty DataFrame
Columns: [invoice_and_item_number, date, store_number, store_name, address, city, zip_code, store_location, county_number, county, category, category_name, vendor_number, vendor_name, item_number, item_description, pack, bottle_volume_ml, state_bottle_cost, state_bottle_retail, bottles_sold, sale_dollars, volume_sold_liters, volume_sold_gallons, year, month]
Index: []

## DataFrame.columns

In [17]:
df.columns

Index(['invoice_and_item_number', 'date', 'store_number', 'store_name',
       'address', 'city', 'zip_code', 'store_location', 'county_number',
       'county', 'category', 'category_name', 'vendor_number', 'vendor_name',
       'item_number', 'item_description', 'pack', 'bottle_volume_ml',
       'state_bottle_cost', 'state_bottle_retail', 'bottles_sold',
       'sale_dollars', 'volume_sold_liters', 'volume_sold_gallons', 'year',
       'month'],
      dtype='object')

## Column `category_name` Values

In [32]:
df['category'].value_counts().index

Float64Index([1031100.0, 1012100.0, 1011200.0, 1081600.0, 1062400.0, 1031200.0,
              1011100.0, 1081400.0, 1051100.0, 1032100.0, 1022200.0, 1011400.0,
              1081200.0, 1081300.0, 1071100.0, 1062200.0, 1062500.0, 1082000.0,
              1022100.0, 1032200.0, 1082200.0, 1041100.0, 1012200.0, 1052100.0,
              1701100.0, 1012400.0, 1042100.0, 1012300.0, 1011600.0, 1081100.0,
              1062100.0, 1091300.0, 1062300.0, 1081500.0, 1091100.0, 1011300.0,
              1091200.0, 1011500.0, 1901200.0, 1092100.0, 1082100.0, 1011700.0,
              1700000.0, 1041200.0, 1022300.0, 1041300.0, 1091400.0, 1031000.0,
              1090000.0, 1091000.0, 1092000.0, 1042000.0, 1032000.0, 1070000.0,
              1081000.0],
             dtype='float64')

## Column `city` Values

In [51]:
df['city'].value_counts().index

Index(['Des Moines', 'Cedar Rapids', 'Davenport', 'Waterloo', 'Sioux City',
       'West Des Moines', 'Ames', 'Council Bluffs', 'Dubuque', 'Iowa City',
       ...
       'Casey', 'Roland', 'Ossian', 'Melbourne', 'Bussey', 'Stanwood',
       'Fontanelle', 'Blairstown', 'Earling', 'Colorado Springs'],
      dtype='object', length=444)

## Column `bottle_volume_ml` Values

In [19]:
df['bottle_volume_ml'].value_counts().index.sort_values()

Int64Index([  20,   50,  100,  200,  300,  375,  400,  500,  600,  700,  750,
             800,  850,  900,  950, 1000, 1200, 1750, 1800, 1950, 2250, 2400,
            3000, 3500, 4500, 5250, 6000],
           dtype='int64')

## DataFrame.describe()

#TODO detecting and treating outliers on columns

- [x] removed due to under 5% outliers
  `pack`, `state_bottle_cost`, `state_bottle_retail`, `bottles_sold`, `sale_dollars`, `volume_sold_liters`
- [x] not removed due to over 5% outliers and are natural outliers
  `bottle_volume_ml`

In [92]:
def calculateLowerLimitUpperLimit(series: pd.Series):
    skew = series.skew()

    # extreme skewed distribution
    if skew < -1 or skew > 1:
        q1, q3 = series.quantile([.25,.75])
        iqr = q3 - q1
        return q1 - iqr * 3, q3 + iqr * 3
    # skewed distribution
    elif skew < -0.5 or skew > 0.5:
        q1, q3 = series.quantile([.25,.75])
        iqr = q3 - q1
        return q1 - iqr * 1.5, q3 + iqr * 1.5
    # normal distribution
    else:
        mean, std = series.agg(['mean', 'std'])
        return mean - 3 * std, mean + 3 * std

for index, value in df.loc[:,['pack', 'bottle_volume_ml', 'state_bottle_cost', 'state_bottle_retail', 
    'bottles_sold', 'sale_dollars', 'volume_sold_liters']].skew().items():
    lowerLimit, upperLimit = calculateLowerLimitUpperLimit(df.loc[:, str(index)])

    min, max = df.loc[:, str(index)].agg(['min', 'max'])
    if lowerLimit > min or upperLimit < max:
        print("column {0}, lower limit {1} > min {2}, upper limit {3} < max {4}" \
                .format(index, lowerLimit, min, upperLimit, max))

column pack, lower limit -12.0 > min 1, upper limit 30.0 < max 60
column bottle_volume_ml, lower limit 375.0 > min 20, upper limit 1375.0 < max 6000
column state_bottle_cost, lower limit -17.0 > min 0.89, upper limit 35.5 < max 1949.02
column state_bottle_retail, lower limit -25.5 > min 1.34, upper limit 53.25 < max 2923.53
column bottles_sold, lower limit -24.0 > min 1, upper limit 39.0 < max 13200
column sale_dollars, lower limit -307.68 > min 0.0, upper limit 490.74 < max 250932.0
column volume_sold_liters, lower limit -25.5 > min 0.02, upper limit 37.5 < max 13200.0


In [57]:
df[['pack','bottle_volume_ml', 'state_bottle_cost', 'state_bottle_retail', 'bottles_sold', 'sale_dollars', 'volume_sold_liters']].describe()

pack  bottle_volume_ml  state_bottle_cost  \
count  373231.000000     373231.000000      373231.000000   
mean       12.293054        863.893674          10.675001   
std         7.919978        515.690884          10.460707   
min         1.000000         20.000000           0.890000   
25%         6.000000        750.000000           5.500000   
50%        12.000000        750.000000           8.500000   
75%        12.000000       1000.000000          13.000000   
max        60.000000       6000.000000        1949.020000   

       state_bottle_retail   bottles_sold   sale_dollars  volume_sold_liters  
count        373231.000000  373231.000000  373231.000000       373231.000000  
mean             16.014474      10.418441     134.513251            8.260086  
std              15.691036      32.321439     546.079015           34.168744  
min               1.340000       1.000000       0.000000            0.020000  
25%               8.250000       3.000000      34.500000            1.500000  
50%              12.750000       6.000000      76.560000            4.500000  
75%              19.500000      12.000000     148.560000           10.500000  
max            2923.530000   13200.000000  250932.000000        13200.000000

## DataFrame.info()

#TODO handle null rows on columns

range index: 373231 entries
- [x] remove null due to categorical data
  - `city`            372256
  - `zip_code`        372256
  - `category`        373094
  - `category_name`   373094

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 373231 entries, 0 to 373230
Data columns (total 26 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   invoice_and_item_number  373231 non-null  object 
 1   date                     373231 non-null  object 
 2   store_number             373231 non-null  int64  
 3   store_name               373231 non-null  object 
 4   address                  372256 non-null  object 
 5   city                     372256 non-null  object 
 6   zip_code                 372256 non-null  float64
 7   store_location           335493 non-null  object 
 8   county_number            372254 non-null  float64
 9   county                   372256 non-null  object 
 10  category                 373094 non-null  float64
 11  category_name            373094 non-null  object 
 12  vendor_number            373231 non-null  float64
 13  vendor_name              373231 non-null  object 
 14  item

# Data Cleaning

In [47]:
df2 = df.sort_values('date').reset_index(drop=True).copy()

In [48]:
df3 = df2.drop(['store_number', 'store_name', 'address', 'store_location', 
                'vendor_number', 'vendor_name', 'item_number', 'item_description', 
                'county_number', 'county', 'volume_sold_gallons', 'year', 'month'], 
                axis='columns') \
         .copy()

## Outliers Detection and Treatment

- natural outliers on columns

  outliers under 5%
  
  `pack`
  
  outliers over 5%

  `bottle_volume_ml`

- not natural outliers on columns

  outliers under 5%

  `state_bottle_cost`, `state_bottle_retail`, `bottles_sold`, `sale_dollars`, `volume_sold_liters`

In [236]:
for index, value in df3.loc[:, ['pack', 'bottle_volume_ml', 'state_bottle_cost', 'state_bottle_retail', 
    'bottles_sold', 'sale_dollars', 'volume_sold_liters']].items():
    lowerLimit, upperLimit = calculateLowerLimitUpperLimit(value)
    outliersPercentage = len(value.loc[(value < lowerLimit) | (value > upperLimit)]) \
                         / len(value)
    print("column {0}, skew {1:.2f}, outliers percentage {2:.2f}%".format(str(index), value.skew(), outliersPercentage * 100))
    print("lower limit {0[0]}, upper limit {0[1]}".format(calculateLowerLimitUpperLimit(value)))
    display(value.value_counts().sort_index(ascending=False))

column pack, skew 2.72, outliers percentage 2.87%
lower limit -12.0, upper limit 30.0


60        18
48      9893
44       807
30         6
28         1
24     36705
20         4
15        54
12    204798
10      8100
9         33
8       1637
6     105226
5        602
4        700
3        448
2          1
1       4198
Name: pack, dtype: int64

column bottle_volume_ml, skew 0.93, outliers percentage 29.22%
lower limit 375.0, upper limit 1375.0


6000        16
5250         1
4500        26
3500        45
3000      1643
2400       132
2250        38
1950         8
1800       107
1750     69668
1200       194
1000     41404
950         14
900         41
850          6
800        263
750     168465
700          1
600       3213
500       9092
400         34
375      41446
300       1301
200      12879
100       3823
50       19341
20          30
Name: bottle_volume_ml, dtype: int64

column state_bottle_cost, skew 49.36, outliers percentage 1.12%
lower limit -17.0, upper limit 35.5


1949.02       1
1599.19       3
419.99        1
390.00        7
314.99        2
           ... 
1.00        227
0.97         73
0.95        245
0.90       2056
0.89        369
Name: state_bottle_cost, Length: 1133, dtype: int64

column state_bottle_retail, skew 49.36, outliers percentage 1.12%
lower limit -25.5, upper limit 53.25


2923.53       1
2398.79       3
629.99        1
585.00        7
472.49        2
           ... 
1.50        227
1.46         73
1.43        245
1.35       2056
1.34        369
Name: state_bottle_retail, Length: 1135, dtype: int64

column bottles_sold, skew 191.71, outliers percentage 3.45%
lower limit -24.0, upper limit 39.0


13200        1
2430         1
2304         1
2100         1
1980         1
         ...  
5         5033
4        19781
3        34862
2        48105
1        42394
Name: bottles_sold, Length: 157, dtype: int64

column sale_dollars, skew 267.76, outliers percentage 2.73%
lower limit -307.68, upper limit 490.74


250932.00     1
36154.80      1
30399.30      1
27181.44      2
26271.00      1
             ..
1.50          4
1.46          4
1.35         13
1.34          3
0.00         79
Name: sale_dollars, Length: 6644, dtype: int64

column volume_sold_liters, skew 178.64, outliers percentage 1.76%
lower limit -25.5, upper limit 37.5


13200.00       1
4252.50        1
4032.00        1
3675.00        1
2688.00        1
            ... 
0.08           1
0.06           2
0.05        6536
0.04           8
0.02           6
Name: volume_sold_liters, Length: 313, dtype: int64

### Ask Instructor

#TODO ask instructor

reference: https://www.quora.com/Is-it-bad-to-delete-statistical-outliers

- [ ] Is it true in Machine Learning, it is always a requirement to remove the outliers?
  - If so, should we remove natural outliers or do mean/median imputation in column `bottle_volume_ml`?

In [232]:
df4 = df3.copy()

# outliers percentage under 5% and not natural outliers
for index, value in df4.loc[:, ['state_bottle_cost', 'state_bottle_retail', 
                                'bottles_sold', 'sale_dollars', 'volume_sold_liters']] \
                        .items():
    lowerLimit, upperLimit = calculateLowerLimitUpperLimit(value)
    df4 = df4.drop(df4.loc[(value < lowerLimit) | (value > upperLimit),str(index)].index)

## Null Treatment

### Ask Instructor

#TODO ask instructor

- [ ] should we remove rows based on condition (a column with categorical data is null)?

In [181]:
for index, value in df4.loc[:,['city','zip_code','category','category_name']].items():
    print("column {0}, null percentage {1:.2f}%".format(index, len(value.loc[value.isna()]) / len(value) * 100))

column city, null percentage 0.25%
column zip_code, null percentage 0.25%
column category, null percentage 0.03%
column category_name, null percentage 0.03%


In [183]:
df5 = df4.copy()

# categorical data, we will drop them.
for index, value in df5.loc[:, ['city', 'zip_code', 'category', 'category_name']].items():
    df5 = df5.drop(df5.loc[value.isna()].index)

In [238]:
df5.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 349528 entries, 0 to 373229
Data columns (total 13 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   invoice_and_item_number  349528 non-null  object 
 1   date                     349528 non-null  object 
 2   city                     349528 non-null  object 
 3   zip_code                 349528 non-null  float64
 4   category                 349528 non-null  float64
 5   category_name            349528 non-null  object 
 6   pack                     349528 non-null  int64  
 7   bottle_volume_ml         349528 non-null  int64  
 8   state_bottle_cost        349528 non-null  float64
 9   state_bottle_retail      349528 non-null  float64
 10  bottles_sold             349528 non-null  int64  
 11  sale_dollars             349528 non-null  float64
 12  volume_sold_liters       349528 non-null  float64
dtypes: float64(6), int64(3), object(4)
memory usage: 37.3+ MB


In [240]:
df6 = df5.reset_index()

# Analisis dan Perhitungan

## Statistik Deskriptif

## Statistik Inferensial

# Pengambilan Keputusan